File: Sample Apps/Java Custom Attribute/src/OtelApi/repository/UserRepository.java
package OtelApi.repository;

import OtelApi.db.DatabaseUtil;
import OtelApi.model.User;

import java.sql.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public class UserRepository {

    public User create(User user) throws SQLException {
        String sql = "INSERT INTO users (username, email, full_name, age, salary, is_active, tags) VALUES (?, ?, ?, ?, ?, ?, ?)";

        try (Connection conn = DatabaseUtil.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, user.getUsername());
            stmt.setString(2, user.getEmail());
            stmt.setString(3, user.getFullName());
            stmt.setInt(4, user.getAge());
            stmt.setDouble(5, user.getSalary());
            stmt.setBoolean(6, user.isActive());
            stmt.setString(7, tagsToString(user.getTags()));

            stmt.executeUpdate();
            return user;
        }
    }

    public Optional<User> findByUsername(String username) throws SQLException {
        String sql = "SELECT username, email, full_name, age, salary, is_active, tags FROM users WHERE username = ?";

        try (Connection conn = DatabaseUtil.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, username);

            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return Optional.of(mapRowToUser(rs));
                }
            }
        }
        return Optional.empty();
    }

    public List<User> findAll() throws SQLException {
        String sql = "SELECT username, email, full_name, age, salary, is_active, tags FROM users";
        List<User> users = new ArrayList<>();

        try (Connection conn = DatabaseUtil.getConnection();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {

            while (rs.next()) {
                users.add(mapRowToUser(rs));
            }
        }
        return users;
    }

    public User update(User user) throws SQLException {
        String sql = "UPDATE users SET email = ?, full_name = ?, age = ?, salary = ?, is_active = ?, tags = ? WHERE username = ?";

        try (Connection conn = DatabaseUtil.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, user.getEmail());
            stmt.setString(2, user.getFullName());
            stmt.setInt(3, user.getAge());
            stmt.setDouble(4, user.getSalary());
            stmt.setBoolean(5, user.isActive());
            stmt.setString(6, tagsToString(user.getTags()));
            stmt.setString(7, user.getUsername());

            int rowsAffected = stmt.executeUpdate();
            if (rowsAffected == 0) {
                throw new SQLException("User not found: " + user.getUsername());
            }
            return user;
        }
    }

    public boolean delete(String username) throws SQLException {
        String sql = "DELETE FROM users WHERE username = ?";

        try (Connection conn = DatabaseUtil.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, username);
            int rowsAffected = stmt.executeUpdate();
            return rowsAffected > 0;
        }
    }

    private User mapRowToUser(ResultSet rs) throws SQLException {
        return new User(
            rs.getString("username"),
            rs.getString("email"),
            rs.getString("full_name"),
            rs.getInt("age"),
            rs.getDouble("salary"),
            rs.getBoolean("is_active"),
            stringToTags(rs.getString("tags"))
        );
    }

    // Convert List<String> to comma-separated string for storage
    private String tagsToString(List<String> tags) {
        if (tags == null || tags.isEmpty()) {
            return "";
        }
        return String.join(",", tags);
    }

    // Convert comma-separated string back to List<String>
    private List<String> stringToTags(String tagsStr) {
        if (tagsStr == null || tagsStr.trim().isEmpty()) {
            return new ArrayList<>();
        }
        return new ArrayList<>(Arrays.asList(tagsStr.split(",")));
    }
}

File: Sample Apps/PHP Custom Attribute/oteltracer/src/Repositories/UserRepository.php
<?php

namespace App\Repositories;

use App\Models\User;
use PDO;
use PDOException;

class UserRepository
{
    private PDO $db;
    private const TABLE_NAME = 'php_user_tbl';

    public function __construct(PDO $db)
    {
        $this->db = $db;
    }

    /**
     * Get all users with optional filters
     */
    public function findAll(array $filters = []): array
    {
        $query = "SELECT * FROM " . self::TABLE_NAME;
        $conditions = [];
        $params = [];

        if (isset($filters['is_active'])) {
            $conditions[] = "is_active = :is_active";
            $params[':is_active'] = $filters['is_active'];
        }

        if (!empty($conditions)) {
            $query .= " WHERE " . implode(' AND ', $conditions);
        }

        $query .= " ORDER BY id ASC";

        $stmt = $this->db->prepare($query);
        $stmt->execute($params);

        $users = [];
        while ($row = $stmt->fetch()) {
            // Convert PostgreSQL array to PHP array
            if (isset($row['tags']) && is_string($row['tags'])) {
                $row['tags'] = $this->parsePostgresArray($row['tags']);
            }
            $users[] = new User($row);
        }

        return $users;
    }

    /**
     * Find user by ID
     */
    public function findById(int $id): ?User
    {
        $query = "SELECT * FROM " . self::TABLE_NAME . " WHERE id = :id";
        $stmt = $this->db->prepare($query);
        $stmt->execute([':id' => $id]);

        $row = $stmt->fetch();
        if (!$row) {
            return null;
        }

        if (isset($row['tags']) && is_string($row['tags'])) {
            $row['tags'] = $this->parsePostgresArray($row['tags']);
        }

        return new User($row);
    }

    /**
     * Find user by username
     */
    public function findByUsername(string $username): ?User
    {
        $query = "SELECT * FROM " . self::TABLE_NAME . " WHERE username = :username";
        $stmt = $this->db->prepare($query);
        $stmt->execute([':username' => $username]);

        $row = $stmt->fetch();
        if (!$row) {
            return null;
        }

        if (isset($row['tags']) && is_string($row['tags'])) {
            $row['tags'] = $this->parsePostgresArray($row['tags']);
        }

        return new User($row);
    }

    /**
     * Create a new user
     */
    public function create(User $user): User
    {
        $query = "INSERT INTO " . self::TABLE_NAME . "
                  (username, email, full_name, age, salary, is_active, tags)
                  VALUES (:username, :email, :full_name, :age, :salary, :is_active, :tags)
                  RETURNING id, created_at, updated_at";

        $stmt = $this->db->prepare($query);
        $stmt->execute([
            ':username' => $user->getUsername(),
            ':email' => $user->getEmail(),
            ':full_name' => $user->getFullName(),
            ':age' => $user->getAge(),
            ':salary' => $user->getSalary(),
            ':is_active' => $user->isActive() ? 'true' : 'false',
            ':tags' => $this->arrayToPostgresArray($user->getTags()),
        ]);

        $result = $stmt->fetch();
        $user->setId($result['id']);
        $user->setCreatedAt($result['created_at']);
        $user->setUpdatedAt($result['updated_at']);

        return $user;
    }

    /**
     * Update an existing user
     */
    public function update(int $id, User $user): ?User
    {

        $query = "UPDATE " . self::TABLE_NAME . " 
                  SET username = :username, email = :email, full_name = :full_name, 
                      age = :age, salary = :salary, is_active = :is_active, tags = :tags
                  WHERE id = :id
                  RETURNING id, created_at, updated_at";

        $stmt = $this->db->prepare($query);
        $stmt->execute([
            ':id' => $id,
            ':username' => $user->getUsername(),
            ':email' => $user->getEmail(),
            ':full_name' => $user->getFullName(),
            ':age' => $user->getAge(),
            ':salary' => $user->getSalary(),
            ':is_active' => $user->isActive() ? 'true' : 'false',
            ':tags' => $this->arrayToPostgresArray($user->getTags()),
        ]);

        $result = $stmt->fetch();
        if (!$result) {
            return null;
        }

        $user->setId($result['id']);
        $user->setCreatedAt($result['created_at']);
        $user->setUpdatedAt($result['updated_at']);

        return $user;
    }

    /**
     * Delete a user
     */
    public function delete(int $id): bool
    {
        $query = "DELETE FROM " . self::TABLE_NAME . " WHERE id = :id";
        $stmt = $this->db->prepare($query);
        $result = $stmt->execute([':id' => $id]);

        return $result && $stmt->rowCount() > 0;
    }

    /**
     * Convert PHP array to PostgreSQL array format
     */
    private function arrayToPostgresArray(array $array): string
    {
        if (empty($array)) {
            return '{}';
        }
        return '{' . implode(',', array_map(function($item) {
            return '"' . str_replace('"', '\\"', $item) . '"';
        }, $array)) . '}';
    }

    /**
     * Parse PostgreSQL array to PHP array
     */
    private function parsePostgresArray(string $pgArray): array
    {
        if ($pgArray === '{}' || empty($pgArray)) {
            return [];
        }

        // Remove curly braces
        $pgArray = trim($pgArray, '{}');

        // Split by comma and clean up quotes
        $items = explode(',', $pgArray);
        return array_map(function($item) {
            return trim($item, '"');
        }, $items);
    }
}

